### PR TITLE
Update tests to test_run_mpi_pi verifies the tasks run concurrently

### LINKED ISSUE #46: test_run_mpi_pi doesn't verify the tasks run concurrently
The `test_run_mpi_pi` test in `test_parsl_flux_mpi_hello.py` ensures that two runs of the `mpi_pi.exe` program do not share nodes.  However, the test does not actually verify that the tests run concurrently.

The output of the tasks is written in `parsl_flux_mpi_pi1_run.out` and `parsl_flux_mpi_pi2_run.out`.  This output includes time stamps for the start and end times of the run.  In order to prove that the two tests run concurrently, two additional assertions are needed.  We need to guarantee that the start time of `pi1` is not greater than the end time of `pi2` and that the start time of `pi2` is not greater than the end time of `pi1`.

An additional problem is that the time stamps only include `HH:MM:SS`.  If the execution of the code spans midnight, then the validation of concurrent execution times will not work.  The Fortran code needs to be updated to write out the full time stamp such that it includes the date as well as the time.

### PR DESCRIPTION
**Pull request by NaureenBharwaniNOAA**
The `test_run_mpi_pi` test in `test_parsl_flux_mpi_hello.py` ensures that two runs of the `mpi_pi.exe` program do not share nodes.  The expansion of the tests verifies that the tests run concurrently.

Confirming in `parsl_flux_mpi_pi1_run.out` and `parsl_flux_mpi_pi2_run.out`, the output has a time stamp that has been updated for the start and end times of the run from a `HH:MM:SS` format to a `YYYY:MM:DD:HH:MM:SS` format. The Fortran code has been updated to reflect these changes. This update was needed to prove that the two tests run concurrently, including two additional assertions.  

Note: We needed to guarantee that the start time of `pi1` is not greater than the end time of `pi2` and that the start time of `pi2` is not greater than the end time of `pi1`.

Currently:

- Time stamp has been updated in the appropriate place to account for this

Still needs to be done:
- Guarantee that the start time of `pi1` is not greater than the end time of `pi2` and that the start time of `pi2` is not greater than the end time of `pi1`.
- Additional two additional assertions to confirm the two tests run concurrently. 

Closes #46.

### CODE REVIEW

File: tests/mpi_pi.f90
Comment (NaureenBharwaniNOAA) [inline tests/mpi_pi.f90:33]: Currently this is what the output from the tests look like for the Start Time:
`Start Time = **: 1:04
:17:26:23`

The year shows up in `**` format, and the tests pass correctly.

File: tests/mpi_pi.f90
Comment (NaureenBharwaniNOAA) [inline tests/mpi_pi.f90:70]: Currently this is what the output from the tests look like for the End Time:
`End Time = **: 1:04
:17:27:02`

The year shows up in `**` format, and the tests pass correctly.

**Comment by NaureenBharwaniNOAA** [comment]
Used this source to update the Fortran time stamp as needed: 
https://faculty.cs.niu.edu/~hutchins/csci297p2/webpages/datetime.htm#:~:text=Suppose%20a%20FORTRAN%20program%20needs,Values%2C%20all%20used%20for%20output.

File: tests/mpi_pi.f90
Comment (christopherwharrop-noaa) [reply tests/mpi_pi.f90:33]: You are definitely on the right track.  You will need to update the string format in the `write` to accommodate the additional fields for the date stamp.  Right now, it is `(A,I2,A,I2,A,I2.2)` but you will need three additional integers and some other characters for the formatting.  The `**` is because the format for the year is `I2` instead of `I4`.  Also, the ` 1` instead of `01` is because that field is `I2` instead of `I2.2`.  I would shoot for having the date stamp in a format that will be easy to parse on the Python side.  Maybe something like: `dd/mm/yyy HH:MM:SS` or some other standardized string-based date format.

File: tests/mpi_pi.f90
Comment (NaureenBharwaniNOAA) [reply tests/mpi_pi.f90:33]: Still trying to figure out where the first 3 numbers of the year aren't showing up, but the format looks much better now.

Comment (christopherwharrop-noaa) [reply tests/mpi_pi.f90:33]: For the year, I think you need `I4` for a 4-digit year.  The `I2.2` is for two-digit numbers that may start with 0.

Comment (NaureenBharwaniNOAA) [reply tests/mpi_pi.f90:33]: I believe that it is in the correct place, `(A,I2.2,A,I2.2,A,I4,A,I2,A,I2,A,I2.2` since the first is `I2.2` for the day -- I left this with the `.2` for the first 9 days of the month format, the second is `I2.2` is for the month and the third is `I4` for the year. 

I'll research more into Fortran and how to write these values and what order they should appear in.

Comment (christopherwharrop-noaa) [reply tests/mpi_pi.f90:33]: You seem to have `timeValues(3), "/", timeValues(2), "/", timeValues(3)` in the code.  The last one needs to be something other than 3.  It's printing the day again with leading spaces... `   4`

Comment (christopherwharrop-noaa) [reply tests/mpi_pi.f90:33]: I think my initial code was bad.  You'll want `2.2` for all of the 2-digit fields because any of them could start with zero.  So, the hour and minute need them too.

Comment (NaureenBharwaniNOAA) [reply tests/mpi_pi.f90:33]: I see what you mean now, I was looking at the wrong thing initially. That bug would have been much easier to see if the day had not been the 4th. I updated that and see it's printing correctly now as `Start Time = 05/01/2024 19:27:26`. I'll push that shortly and incorporate your other comment as well.

**Comment by NaureenBharwaniNOAA** [comment]
I've been failing the tests with this message `Error: Error: No public SSH keys registered with NaureenBharwaniNOAA's GitHub profile`. I can try adding an SSH key to my profile, not sure if I have that set-up, but I definitely have a personal access token set-up and usually clone the `HTTPS` version on a GitHub repo.

**Comment by christopherwharrop-noaa** [comment]
@NaureenBharwaniNOAA - The ssh key message isn't the failure.  It's having a problem with the run hello and the mpi_pi tests.  Not sure what the issue is.  I assume tests pass locally?

**Comment by NaureenBharwaniNOAA** [comment]
> @NaureenBharwaniNOAA - The ssh key message isn't the failure. It's having a problem with the run hello and the mpi_pi tests. Not sure what the issue is. I assume tests pass locally?

Yes, the tests do pass locally for me just fine. I was going to begin worked on the additional assert statements for the second half of this ticket. Do you think that it should be fine? Or should I hold off?

**Comment by christopherwharrop-noaa** [comment]
Keep working on it.  I'll clone your branch and see if I can figure out what is going on.

File: tests/test_parsl_flux_mpi_hello.py
Comment (NaureenBharwaniNOAA) [inline tests/test_parsl_flux_mpi_hello.py:143]: Interesting, after I did a `git pull` and merged main into my bugfix branch, I did a `git commit` and pushed here and this showed up.

Comment (NaureenBharwaniNOAA) [reply tests/test_parsl_flux_mpi_hello.py:143]: I'll discard these lines since they seem to be replicated from the above ones

File: tests/test_parsl_flux_mpi_hello.py
Comment (NaureenBharwaniNOAA) [inline tests/test_parsl_flux_mpi_hello.py:209]: I was thinking there could be a better way to DRY this out since it is the same thing, but using each `*.out` file.

File: tests/test_parsl_flux_mpi_hello.py
Comment (NaureenBharwaniNOAA) [inline tests/test_parsl_flux_mpi_hello.py:209]: **I was going to try to do something here like:**
```
   with open("parsl_flux_mpi_pi1_run.out", "r") as pi1, ("parsl_flux_mpi_pi2_run.out", "r") as pi2: 
        for pi1_line, pi2_line in pi1, pi2:
            if re.match(r"Start Time ", pi1_line), re.match(r"Start Time ", pi2_line):
                pi1_line = pi1_line.strip().lstrip("Start Time = ")               
                pi2_line = pi2_line.strip().lstrip("Start Time = ")

                pi1_start_time = dt.strptime(pi1_line, "%d/%m/%Y %H:%M:%S")
                pi2_start_time = dt.strptime(pi2_line, "%d/%m/%Y %H:%M:%S")

            if re.match(r"End Time ", pi1_line), re.match(r"End Time ", pi2_line):
                pi1_line = pi1_line.strip().lstrip("End Time = ")               
                pi2_line = pi2_line.strip().lstrip("End Time = ")

                pi1_start_time = dt.strptime(pi1_line, "%d/%m/%Y %H:%M:%S")
                pi2_start_time = dt.strptime(pi2_line, "%d/%m/%Y %H:%M:%S")

    assert pi1_start_time < pi2_end_time and pi2_start_time < pi1_end_time
```

If you have any suggestions to DRY this out, I'm willing to incorporate it.

File: tests/test_parsl_flux_mpi_hello.py
Comment (NaureenBharwaniNOAA) [inline tests/test_parsl_flux_mpi_hello.py:1]: I can also try to add the lines in the code block before so we don't have to reopen the files -- 
I added plus signs next to the additional lines
```
    # Extract the hostnames used by pi1
    with open("parsl_flux_mpi_pi1_run.out", "r") as f:
        pi1_hosts=[]
        for line in f:
            if re.match(r"Host ", line):
                pi1_hosts.append(line.split()[1])

           + if re.match(r"Start Time ", line):
           +   line = line.strip().lstrip("Start Time = ")
           +   pi1_start_time = dt.strptime(line, "%d/%m/%Y %H:%M:%S")
           + if re.match(r"End Time ", line):
           +   line = line.strip().lstrip("End Time = ")
           +   pi1_end_time = dt.strptime(line, "%d/%m/%Y %H:%M:%S")

    # Extract the hostnames used by pi2
    with open("parsl_flux_mpi_pi2_run.out", "r") as f:
        pi2_hosts=[]
        for line in f:
            if re.match(r"Host ", line):
                pi2_hosts.append(line.split()[1])

           + if re.match(r"Start Time ", line):
           +   line = line.strip().lstrip("Start Time = ")
           +   pi2_start_time = dt.strptime(line, "%d/%m/%Y %H:%M:%S")
           + if re.match(r"End Time ", line):
           +   line = line.strip().lstrip("End Time = ")
           +   pi2_end_time = dt.strptime(line, "%d/%m/%Y %H:%M:%S")


    # Verify each pi test ran on a different set of nodes
    assert set(pi1_hosts).intersection(pi2_hosts) == set()
    + assert pi1_start_time < pi2_end_time and pi2_start_time < pi1_end_time
```

**Comment by christopherwharrop-noaa** [comment]
Ok... How about this:

```
start_time=[]
end_time=[]
for i in (1,2):
    with open('''parsl_flux_mpi_pi{}_run.out'''.format(i), "r") as pi: 
        for line in pi:
            if re.match(r"Start Time ", line):
                line = line.strip().lstrip("Start Time = ")
                start_time.append(dt.strptime(line, "%d/%m/%Y %H:%M:%S"))
            if re.match(r"End Time ", line):
                line = line.strip().lstrip("End Time = ")
                end_time.append(dt.strptime(line, "%d/%m/%Y %H:%M:%S"))
assert start_time[0] < end_time[1] and start_time[1] < end_time[0]
 ```

**Comment by christopherwharrop-noaa** [comment]
You could probably DRY it out even more....
```
times={}
for i in (1,2):
    with open('''parsl_flux_mpi_pi{}_run.out'''.format(i), "r") as pi: 
        for line in pi:
            for tstr in ("Start Time", End Time"):
                times[tstr]=[]
                if re.match(r'''{} '''.format(tstr), line):
                    line = line.strip().lstrip('''{} = '''.format(tstr))
                    times[tstr].append(dt.strptime(line, "%d/%m/%Y %H:%M:%S"))
assert times["Start Time"][0] < times["End Time"][1] and times["Start Time"][1] < times["End Time"][0]
```

**Review by christopherwharrop-noaa — APPROVED** [review]
Great work!  Thank you for this important fix to the tests.

**Comment by christopherwharrop-noaa** [comment]
Please go ahead and merge this when you have a moment.

**Comment by christopherwharrop-noaa** [comment]
BTW - Your solution for DRYing the code was much better than mine!

**Pull request merged**
